### PR TITLE
Set the image pull policy

### PR DESCRIPTION
**Pull request by giovannibianco**
Set the pull policy in the orchestrator

### CODE REVIEW

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 1]
```diff
@@ -47,13 +47,36 @@ def __init__(self):
...
+        self.namespace = \
+            self.get_nuvlaedge_project_name(util.default_project_name)
+        self.job_engine_lite_image = \
+            os.getenv('NUVLAEDGE_JOB_ENGINE_LITE_IMAGE') or self.current_image
         self.host_node_ip = os.getenv('MY_HOST_NODE_IP')
         self.host_node_name = os.getenv('MY_HOST_NODE_NAME')
-        self.vpn_client_component = os.getenv('NUVLAEDGE_VPN_COMPONENT_NAME', 'vpn-client')
+        self.vpn_client_component = \
+            os.getenv('NUVLAEDGE_VPN_COMPONENT_NAME', 'vpn-client')
+        self.set_image_pull_policy = self.checked_image_pull_policy()
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:58]: This is specific for `job` so this variable should represent that.
```python
        self.job_image_pull_policy = self.checked_image_pull_policy()
```
.
The function/method should also represent that 
```suggestion
        self.job_image_pull_policy = self.get_job_image_pull_policy()
```
or maybe better, the `getenv` should be extracted to make the function/method generic
```suggestion
        job_img_pull_policy = os.getenv('JOB_IMAGE_PULL_POLICY')
        self.job_image_pull_policy = self.get_image_pull_policy(job_img_pull_policy)
```

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 2]
```diff
@@ -47,13 +47,36 @@ def __init__(self):
...
+
+        valid_pull_policies = ["Always","IfNotPresent","Never"]
+
+        image_pull_policy = os.getenv('JOB_ENGINE_IMAGE_PULL_POLICY')
+
+        if image_pull_policy in valid_pull_policies:
+            return image_pull_policy
+
+        log.info(f"The image pull policy was set to an invalid string: \
+            {image_pull_policy}")
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:75]: I think this should be a warning then:
```suggestion
        log.warning("The image pull policy was set to an invalid string: "
                    f"{image_pull_policy}")
```

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 3]
```diff
@@ -47,13 +47,36 @@ def __init__(self):
...
+        self.set_image_pull_policy = self.checked_image_pull_policy()
         self.data_gateway_name = f"data-gateway.{self.namespace}"
 
+    def checked_image_pull_policy(self):
+        '''
+        Check if the image pull policy is valid
+        If not, return a sane value of IfNotPresent
+        '''
+
+        valid_pull_policies = ["Always","IfNotPresent","Never"]
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:67]:
```suggestion
        valid_pull_policies = ["Always", "IfNotPresent", "Never"]
```

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 1 above)
Comment (schaubl) [reply nuvlaedge/agent/orchestrator/kubernetes.py:58]: And as you can see in my second "suggested change", I would rename the `JOB_ENGINE_IMAGE_PULL_POLICY` to `JOB_IMAGE_PULL_POLICY` to be shorter and match the format we have on _docker-compose.yml_ file for docker NE.

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 4]
```diff
@@ -47,13 +47,36 @@ def __init__(self):
...
+
+        image_pull_policy = os.getenv('JOB_ENGINE_IMAGE_PULL_POLICY')
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:69]: As mentioned in a comment above, I would extract this to make this method generic and static.
```suggestion
    @staticmethod
    def get_image_pull_policy(self, image_pull_policy=DEFAULT_IMAGE_PULL_POLICY):
        '''
        Check if the image pull policy is valid
        If not, return a sane value of IfNotPresent
        '''

        valid_pull_policies = ["Always", "IfNotPresent", "Never"]

```
`DEFAULT_IMAGE_PULL_POLICY` will have to be defined on the top this class.

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 5]
```diff
@@ -500,7 +523,8 @@ def _to_k8s_obj_name(name: str) -> str:
     def _container_def(image, name,
                        volume_mounts: List[client.V1VolumeMount] | None,
                        command: str = None,
-                       args: str = None) -> client.V1Container:
+                       image_pull_policy: str = None,
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:526]: What happen if it is `None` ? 
Does it use a default ? If yes, what is the default and where does it come from (Kubernetes Python library, Kubernetes "daemon" configuration/default ?

Except if it uses a default (on the daemon) based on user configuration, I would specify a default here.

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 6]
```diff
@@ -517,7 +541,8 @@ def parse_cmd_args(cmd, arg):
                                   name=name,
                                   command=command,
                                   volume_mounts=volume_mounts,
-                                  args=args)
+                                  image_pull_policy=image_pull_policy,
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:544]: If the method/function `checked_image_pull_policy()`is made generic as suggested, I would use it here.
```suggestion
                                  image_pull_policy=self.get_image_pull_policy(image_pull_policy),
```

**Review by schaubl — CHANGES_REQUESTED** [review]
As far a I can tell this PR seems to have no effect because `checked_image_pull_policy()` is only called to define `self.set_image_pull_policy` which is never used.

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 1 above)
Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py]: I've changed the name to reflect the job image policy.
Changed the call to the checked_image_pull_policy
Added the default constant

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 2 above)
Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py]: Changed to a warning. Also the formatting.

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 3 above)
Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py]: Spacing done.

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 4 above)
Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py]: Changed the internals of the function.
Changed the name.

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 5 above)
Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py]: I have added the default using the DEFAULT constant at the top of the class.

File: nuvlaedge/agent/orchestrator/kubernetes.py (hunk 6 above)
Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py:546]: Hmm... I cannot use the static method inside another static method.
I use the "image_pull_policy=self.job_image_pull_policy," as the check is already performed for this variable.

**Comment by giovannibianco** [comment]
All changes have been made and the hepm upgrade with all possible values plus a wrong value have been tested to work OK.

File: nuvlaedge/agent/orchestrator/kubernetes.py [hunk 7]
```diff
@@ -37,23 +37,48 @@ class KubernetesClient(COEClient):
 
     WAIT_SLEEP_SEC = 2
 
-    # FIXME: This needs to be parametrised.
+    # FIXME: This needs to be parameterised.
     NE_DB_ROOT_HOSTPATH = '/var/lib/nuvlaedge'
     NE_DB_CONTAINER_PATH = str(FILE_NAMES.root_fs)
 
+    DEFAULT_IMAGE_PULL_POLICY = "IfNotPresent"
```
Comment (schaubl) [inline nuvlaedge/agent/orchestrator/kubernetes.py:44]: Should we not put "Always" here as well ?

Comment (giovannibianco) [reply nuvlaedge/agent/orchestrator/kubernetes.py]: Ah rats... Yes, you're right. I've changed it and pushed. 
I suppose having "Always" makes me a bit nervous...

**Comment by giovannibianco** [comment]
Ready to merge?

**Pull request merged**